File: leetcode/876.cpp
/*
Question:
Given the head of a singly linked list, return the middle node of the linked list.
If there are two middle nodes, return the second middle node.
*/

// Approach:
// If two pointers are made and one is made to travel at double speed, 
// then by the time the fast pointer reaches the end of the LL, the slow one would be at the middle
// 1) We make a fast and slow pointer and set it to the head of the LL
// 2) We advance fast by 2 and slow by 1 steps while fast and fast->next are valid
// 3) We return slow at the end

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode *slow = head, *fast = head;
        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
        }
        return slow;
    }
};

File: leetcode/876.java
/*
Question:
Given the head of a singly linked list, return the middle node of the linked list.

If there are two middle nodes, return the second middle node.
*/

// Approach:
// Make two pointers, slow and fast. 
// Slow moves one step at a time and fast moves two steps at a time. 
// When fast reaches the end, slow will be at the middle of the list.
// So can return slow.

class Solution {
  public ListNode middleNode(ListNode head) {
    // Create the two pointers
    ListNode slow = head, fast = head;

    // Loop until fast reaches the end of the list
    while(fast != null && fast.next != null) {
      // Move slow one step and fast two steps
      fast = fast.next.next;
      slow = slow.next;
    }

    // Return the slow pointer
    return slow;
  }
}
